### DISCUSSION TITLE
Discord OAuth not working, but GitHub OAuth does (SvelteKit & PocketBase JavaScript SDK)

### DISCUSSION THREAD

**Discussion by janniklasdiederich** [pocketbase/pocketbase#2159] [Q&A] [answered]
Hello there!
I'm currently facing a problem where I'm using **SvelteKit** and  the **Pocketbase JavaScript SDK**.
For the website I'm working on I want to use OAuth for authentication, specifically discord, 
but I'm getting the following error message: 
`{"error": "invalid_grant", "error_description": "Code challenge failed."}`

I've already tried to search in similar discussions and have checked that none of those are the error.
For example in #1739 I have already made sure that **it's neither of those**:
> The Create API rule of your auth collection doesn't allow creating new users.
> "Always require email" collection option is enabled (could be changed from the collection settings > Options tab).

I've also made sure that the correct Redirect URL has been set in the discord developer portal (I've set it to `http://localhost:5173/callback`)

As I can't seem to find the answer on my own it would be **greatly appreciated** if someone could **help me** from this awesome community! :)

Here are the relevant parts of my current code:

**src/routes/login/+page.server.ts**
```typescript
import type { PageServerLoad } from './$types';

export type OutputType = { 
    authProviderUrls: Map<string, string>,
    authProviderStates: Map<string, string>
 };

export const load: PageServerLoad<OutputType> = async ({ locals, url }) => {
    const callbackURL = `${url.origin}/callback`;
    const authMethods = await locals.pb?.collection('users').listAuthMethods();
    const authProviderUrls = new Map<string, string>();
    const authProviderStates = new Map<string, string>();

    authMethods.authProviders.forEach((provider) => {
        authProviderUrls.set(provider.name, provider.authUrl + callbackURL);
        authProviderStates.set(provider.name, provider.state);
    });

    return {
        authProviderUrls: authProviderUrls,
        authProviderStates: authProviderStates
    };
};
```

**src/routes/login/+page.server.ts**
```svelte
<script lang="ts">
  import { browser } from '$app/environment';
  import type { PageData } from './$types';

  export let data: PageData;

  function providerSelected(providerName: string, providerUrl: string) {
    if (browser) {
      document.cookie = `state=${data?.authProviderStates.get(providerName)}`;
      document.cookie = `provider=${providerName}`;
    }

    window.location.href = providerUrl;
  }
</script>

{#each [...data.authProviderUrls] as [key, value]}
	<button on:click={() => {providerSelected(key, value) }}>
    {key}
  </button>
{/each}
```

**src/routes/login/+page.server.ts**
```typescript
import { redirect } from '@sveltejs/kit';
import type { RequestEvent, RequestHandler } from './$types';

export const GET: RequestHandler = async ({ locals, url, cookies }: RequestEvent) => {
    const query = new URLSearchParams(url.search);
    const code = query.get('code');
    const state = query.get('state');

    if (!code || !state) {
        throw redirect(303, '/login');
    }

    const authMethods = await locals.pb?.collection('users').listAuthMethods();
    if (!authMethods?.authProviders) {
        throw redirect(303, '/login');
    }

    const provider = authMethods.authProviders.find((p) => p.name === cookies.get('provider'));
    if (!provider) {
        throw redirect(303, '/login');
    }

    const callbackURL = `${url.origin}/callback`;
    const expectedState = cookies.get('state');

    if (expectedState !== state) {
        throw redirect(303, '/login');
    }

    try {
        await locals.pb
            ?.collection('users')
            .authWithOAuth2(provider.name, code, provider.codeVerifier, callbackURL);
    } catch (err) {
        console.log('Error logging in with 0Auth user', err);
    }

    throw redirect(303, '/');
};
```

If there is anything unclear about my code I'll be happy to provide my answers / more context

**Comment by ganigeorgiev** [comment, marked as answer]
In the last code sample you have to use the `provider.codeVerifier` from **before** the redirect. `listAuthMethods()` returns a new `codeChallenge` and `codeVerifier` every time when you call it.

Your existing workflow may work with the GitHub OAuth2 because I think the `codeVerifier` is ignored since they currently don't support the PKCE flow (I'm not sure if the same apply for the enterprise version).

I've also just tested the integration with a Discord OAuth2 app following the plain JS example in https://pocketbase.io/docs/authentication/#web-oauth2-integration and it is working fine for me.

**Reply by janniklasdiederich** [reply]
Thank you so much!
It turns out that I really just had to use the same provider.codeVerifier.. such a stupid mistake on my part. 
Thank your for the help, greatly appreciated.
Cheers!

**Reply by meltinbit** [reply]
Hi! I have the same problem...I kind understood your discussion above but i can get around it. Can you please better explain what you changed in your logic to provide the right codeVerifier? thanks
